### ISSUE TITLE
`Attempt to read after eof` in inverted index

### ISSUE THREAD

**Issue by cangyin** [ClickHouse/ClickHouse#52604]
```log
[chi-datalake-ck-cluster-2-1-0] 2023.07.26 17:29:39.589948 [ 1937 ] {2d370291-66f9-4228-9d5a-3ac0362b6f90} <Error> executeQuery: Code: 32. DB::Exception: Attempt to read after eof. (ATTEMPT_TO_READ_AFTER_EOF) (version 23.5.4.10) (from 127.0.0.1:49598) (in query: ..., Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000e0c3d95 in /usr/bin/clickhouse
1. ? @ 0x000000000875ea58 in /usr/bin/clickhouse
2. DB::throwReadAfterEOF() @ 0x000000000e12d1e3 in /usr/bin/clickhouse
3. ? @ 0x000000000e112e33 in /usr/bin/clickhouse
4. DB::GinIndexStoreDeserializer::readSegmentDictionary(unsigned int) @ 0x000000001451b994 in /usr/bin/clickhouse
5. DB::GinIndexStoreFactory::get(String const&, std::shared_ptr<DB::IDataPartStorage const>) @ 0x000000001451d9d6 in /usr/bin/clickhouse
6. DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::shared_ptr<DB::IMergeTreeIndex const>, std::shared_ptr<DB::IMergeTreeIndexCondition>, std::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, unsigned long&, DB::MarkCache*, DB::UncompressedCache*, Poco::Logger*) @ 0x000000001466a1b6 in /usr/bin/clickhouse
7. ? @ 0x0000000014666f9a in /usr/bin/clickhouse
8. ? @ 0x0000000014671bfb in /usr/bin/clickhouse
9. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000e19be23 in /usr/bin/clickhouse
10. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000e19e615 in /usr/bin/clickhouse
11. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000e197cb4 in /usr/bin/clickhouse
12. ? @ 0x000000000e19d4a1 in /usr/bin/clickhouse
13. ? @ 0x00007f1fb2094609 in ?
14. clone @ 0x00007f1fb1fb9133 in ?


0 rows in set. Elapsed: 0.073 sec.

Received exception from server (version 23.5.4):
Code: 32. DB::Exception: Received from localhost:9000. DB::Exception: Attempt to read after eof. (ATTEMPT_TO_READ_AFTER_EOF)
```

**Comment by cangyin** [comment]
```sh
llvm-addr2line-16 -pafiCse ./clickhouse 0x000000001451b994
0x1451b994: std::__1::shared_ptr<DB::GinSegmentDictionary>::operator->[abi:v15000]() const at shared_ptr.h:815
 (inlined by) DB::GinIndexStoreDeserializer::readSegmentDictionary(unsigned int) at GinIndexStore.cpp:388
```

**Comment by rschu1ze** [comment]
@cangyin Do you have some repro?

**Comment by cangyin** [comment]
hard to reproduce.

this error occurs after upgrading from 22.9 (with tables containing `gin`, now `inverted`, indexes), and the `.gin_post` files did not get updated by `ALTER TABLE MATERIALIZE INDEX`.

```sql
DROP TABLE IF EXISTS test_gin;

CREATE TABLE test_gin
(
    `p` String,
    `body` String,
    INDEX ginidx body TYPE inverted(3, 0.6) GRANULARITY 1,
)
ENGINE = MergeTree
PARTITION BY p
ORDER BY p
SETTINGS min_rows_for_wide_part=1,min_bytes_for_wide_part=1;

INSERT INTO test_gin VALUES ('p1', 'The sun was setting behind the mountains.');
```

**ClickHouse version**
Release Branch v23.5, with patches:
- https://github.com/ClickHouse/ClickHouse/pull/50104
- https://github.com/ClickHouse/ClickHouse/pull/47663 (only applied code block in `collectFilesToSkip`)

start debugging. (yes, debug!)

set breakpoint here:
https://github.com/ClickHouse/ClickHouse/blob/d89e2e6a27746dbb8febd2990d1ed3c23fcf153b/src/Storages/MergeTree/MutateTask.cpp#L549-L554

with patch #47663 applied, the code should actually look like:
```
    for (const auto & index : indices_to_recalc)
    {
        /// Since MinMax index has .idx2 extension, we need to add correct extension.
        files_to_skip.insert(index->getFileName() + index->getSerializedFileExtension());
        files_to_skip.insert(index->getFileName() + mrk_extension);

        /// skip all inverted index files
        if (typeid_cast<const MergeTreeIndexInverted *>(&*index) != nullptr)
        {
            files_to_skip.insert(index->getFileName() + ".gin_dict");
            files_to_skip.insert(index->getFileName() + ".gin_post");
            files_to_skip.insert(index->getFileName() + ".gin_sed");
            files_to_skip.insert(index->getFileName() + ".gin_sid");
        }
    }
```

try 1:
```sql
ALTER TABLE test_gin UPDATE body=body WHERE 1;
```
it steps into the `for` loop, as expected.


try 2:
```sql
ALTER TABLE test_gin MATERIALIZE INDEX ginidx;
```
it does not step into the `for` loop!



In short: the inverted index is not collected as that should be recalculated (`indices_to_recalc`) during execution of `ALTER TABLE MATERIALIZE INDEX`. @amosbird 


Even if the '.gin_*' files is collected to be skipped and recalculated, the error `Attempt to read after eof` still occur. I have to manually remove all the '*.gin_' files and re-MATERIALIZE it, to finally make it right.

**Comment by cangyin** [comment]
I am switching back to branch 22.9 due to personal work schedule.

**Comment by cangyin** [comment]
this code logic:

https://github.com/ClickHouse/ClickHouse/blob/d89e2e6a27746dbb8febd2990d1ed3c23fcf153b/src/Interpreters/MutationsInterpreter.cpp#L862-L874

may ignore any index names from `ALTER TABLE MATERIALIZE INDEX`.
